### PR TITLE
REST API: add flag coverage to flag list endpoint

### PR DESCRIPTION
**Pull request by JerrySentry**
closes https://github.com/codecov/feedback/issues/362

Updates `/api/v2/{service}/{owner_username}/repos/{repo_name}/flags/` endpoint to include the coverage value for each flag in the response object.

Old: `"results": [{"flag_name": string}]`
New: `"results": [{"flag_name": string, "coverage": float}]`

This coverage value comes from the computed report of the head commit of the repo's default branch. If failure to retrieve such value a default value of 0 will be returned.

### Legal Boilerplate

Look, I get it. The entity doing business as "Sentry" was incorporated in the State of Delaware in 2015 as Functional Software, Inc. In 2022 this entity acquired Codecov and as result Sentry is going to need some rights from me in order to utilize my contributions in this PR. So here's the deal: I retain all rights, title and interest in and to my contributions, and by keeping this boilerplate intact I confirm that Sentry can use, modify, copy, and redistribute my contributions, under Sentry's choice of terms.

### CODE REVIEW

File: api/public/v2/flag/views.py [hunk 1]
```diff
@@ -15,11 +21,24 @@ class FlagViewSet(viewsets.GenericViewSet, mixins.ListModelMixin, RepoPropertyMi
...
 
-    def get_queryset(self):
-        return self.repo.flags.all()
+    def get_queryset(self) -> QuerySet:
+        results = [
+            {"flag_name": f.flag_name, "coverage": 0} for f in self.repo.flags.all()
+        ]
+        try:
+            report = self.get_commit().full_report
+            if not report:
```
Comment (ajay-sentry) [inline api/public/v2/flag/views.py:30]: just curious, when would we throw a NotFound vs. return None?

File: api/public/v2/flag/views.py [hunk 2]
```diff
@@ -15,11 +21,24 @@ class FlagViewSet(viewsets.GenericViewSet, mixins.ListModelMixin, RepoPropertyMi
...
+                return results
+        except NotFound:
+            return results
+
+        for i, val in enumerate(results):
+            flag_report = report.filter(flags=[val["flag_name"]])
+            results[i]["coverage"] = flag_report.totals.coverage or 0
```
Comment (ajay-sentry) [inline api/public/v2/flag/views.py:37]: totals will always exist?

File: api/public/v2/tests/test_flag_viewset.py [hunk 3]
```diff
@@ -38,15 +66,64 @@ def _request_flags(self):
...
+        get_repo_permissions.return_value = (True, True)
+
+        res = self._request_flags()
+        assert res.status_code == 200
+        assert res.json() == {
+            "count": 2,
+            "next": None,
+            "previous": None,
+            "results": [
+                {"flag_name": "foo", "coverage": 0},
```
Comment (ajay-sentry) [inline api/public/v2/tests/test_flag_viewset.py:79]: Jw on returning 0 for coverage here vs. None 🤔

File: api/public/v2/flag/views.py (hunk 1 above)
Comment (JerrySentry) [reply api/public/v2/flag/views.py:30]: this line would return the list of flags without the coverage information for each flag if the coverage report of the head commit is not yet processed.
in general imo we should throw 404 if the requested data is not found (eg in the param you have a repo/owner that doesn't exist) so there's nothing to return. And return null, I don't think we would return that as a response to the entire request, rather the response object has some partial data is not available.

File: api/public/v2/flag/views.py (hunk 2 above)
Comment (JerrySentry) [reply api/public/v2/flag/views.py:37]: from what I can tell yes

File: api/public/v2/tests/test_flag_viewset.py (hunk 3 above)
Comment (JerrySentry) [reply api/public/v2/tests/test_flag_viewset.py:79]: yeah I feel none is better

File: api/public/v2/flag/views.py (hunk 1 above)
Comment (ajay-sentry) [reply api/public/v2/flag/views.py:30]: Gotcha! Makes sense to me, thanks for the context

File: api/public/v2/flag/views.py (hunk 2 above)
Comment (ajay-sentry) [reply api/public/v2/flag/views.py:37]: sggg

**Review by ajay-sentry — APPROVED** [review]
looks good 👌

**Pull request merged**